### PR TITLE
For #8791: Replace fenix tab counter with AC version

### PR DESCRIPTION
**Pull request by eliserichards**
For #8791
Depends on https://github.com/mozilla-mobile/android-components/pull/8872

Todo
- [x] add dependencies for ui-tabcounter
   `mozilla_ui_tabcounter = "org.mozilla.components:ui-tabcounter:${Versions.mozilla_android_components}"`
- [ ] ~replace TabCounter~
- [x] replace TabCounterMenu
- [x] replace TabCounterToolbarButton
- [x] ToolbarIntegration
- [x] add metrics in onItemTapped - `metrics.track(Event.TabCounterMenuItemTapped(Event.TabCounterMenuItemTapped.Item.NEW_TAB))`
- [x] replace NewTab(mode) with NewTab/NewPrivateTab
- [x] Tests!
   - [x] TabCounterMenuTest
   - [x] BrowserInteractorTest
   - [x] DefaultBrowserToolbarControllerTest

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by eliserichards** [comment]
This is ready to be tested with https://github.com/mozilla-mobile/android-components/pull/8872 (branch name: `eliserichards:fenix-8791/tab-counter`)

**Comment by eliserichards** [comment]
Private mode 1 tab


Private mode 10+ tabs

File: app/src/main/java/org/mozilla/fenix/components/toolbar/FenixTabCounterToolbarButton.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        showTabs,
+        store,
+        tabCounterMenu
+    ) {
+
+    override fun createView(parent: ViewGroup): View {
+        val view = super.createView(parent)
+
+        view.apply {
+            addOnAttachStateChangeListener(object : View.OnAttachStateChangeListener {
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/FenixTabCounterToolbarButton.kt:42]: This workaround really shouldn't be needed anymore, see https://github.com/mozilla-mobile/android-components/pull/8872#discussion_r516762918

Did you run into problem without this? If so, let's try and fix them.

File: app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserToolbarController.kt [hunk 2]
```diff
@@ -105,6 +106,9 @@ class DefaultBrowserToolbarController(
     override fun handleTabCounterItemInteraction(item: TabCounterMenu.Item) {
         when (item) {
             is TabCounterMenu.Item.CloseTab -> {
+                metrics.track(
+                    Event.TabCounterMenuItemTapped(Event.TabCounterMenuItemTapped.Item.CLOSE_TAB)
+                )
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserToolbarController.kt:111]: Looks like unrelated changes here?

File: app/src/main/java/org/mozilla/fenix/components/toolbar/FenixTabCounterToolbarButton.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+            val tint = when (isPrivate(store)) {
+                true -> {
+                    ContextCompat.getColor(context, R.color.primary_text_default_material_dark)
+                }
+                else -> {
+                    ContextCompat.getColor(context, R.color.mozac_ui_tabcounter_default_tint)
+                }
+            }
+            super.reference.get().tintDrawables(tint)
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/components/toolbar/FenixTabCounterToolbarButton.kt:60]: ^^ With this we have the wrong tint on the private browsing home fragment (without a tab). You can reproduce by removing all private tabs, but I think we can actually remove this logic here as well and apply the tint in `HomeFragment.updateTabCounter`.

```kotlin
private fun updateTabCounter(browserState: BrowserState) {
    val tabCount = if (browsingModeManager.mode.isPrivate) {
        view?.tab_button?.tintDrawables(
            ContextCompat.getColor(requireContext(), mozilla.components.ui.tabcounter.R.color.primary_text_default_material_dark)
        )
        browserState.privateTabs.size
    } else {
        view?.tab_button?.tintDrawables(
            ContextCompat.getColor(requireContext(), mozilla.components.ui.tabcounter.R.color.mozac_ui_tabcounter_default_tint)
        )
        browserState.normalTabs.size
    }

    view?.tab_button?.setCountWithAnimation(tabCount)
    view?.add_tabs_to_collections_button?.isVisible = tabCount > 0
}
```

This way we also don't need to make the weak reference accessible (see my comment on the A-C PR: https://github.com/mozilla-mobile/android-components/pull/8872#discussion_r526154128)

**Review by csadilek — CHANGES_REQUESTED** [review]
Great progress! A few things left, mostly in the A-C PR.

File: app/src/main/java/org/mozilla/fenix/components/toolbar/FenixTabCounterToolbarButton.kt (hunk 1 above)
Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/components/toolbar/FenixTabCounterToolbarButton.kt:42]: Nope, no problems, just a faulty memory ;)

**Comment by eliserichards** [comment]
Todo: 

- [ ] fix tint in private mode
- [ ] long pressing on menu not triggering `menuController.show(anchor = it)`
- [ ] TabCounterToolbarButtonTest in AC

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt [hunk 4]
```diff
@@ -964,8 +978,11 @@ class HomeFragment : Fragment() {
         )
     }
 
+    // TODO use [FenixTabCounterToolbarButton] instead of [TabCounter]: Link to issue
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:981]: Let's file and link to a new ticket here to use `TabCounterToolbarButton` in HomeFragment.

Comment (eliserichards) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:980]: https://github.com/mozilla-mobile/fenix/issues/16792

**Comment by eliserichards** [comment]
## Test cases to confirm
_HomeFragment and BrowserFragment have different implementations, see #16792_

### Home fragment
- [ ] Light mode, normal browsing mode
- [ ] Light mode, private browsing mode
- [ ] Dark mode, normal browsing mode
- [ ] Dark mode, private browsing mode
- [ ] Tab counter spacing around numbers: single digit, double, infinity

### Browser fragment
- [ ] Light mode, normal browsing mode
- [ ] Light mode, private browsing mode
- [ ] Dark mode, normal browsing mode
- [ ] Dark mode, private browsing mode
- [ ] Private mode, close all open tabs
- [ ] Normal mode, close all open tabs
- [ ] Tab counter spacing around numbers: single digit, double, infinity

**Comment by eliserichards** [comment]
Unaligned infinite character: 
image

**Comment by eliserichards** [comment]
## Testing round 2

## Test cases to confirm
_HomeFragment and BrowserFragment have different implementations, see #16792_

### Home fragment
- [x] Light mode, normal browsing mode
- [x] Light mode, private browsing mode
- [x] Dark mode, normal browsing mode
- [x] Dark mode, private browsing mode
- [ ] Tab counter spacing around numbers: 
   ✅ single digit
   ✅ double
   ✅ infinity
- [x] Long press tab counter
  - [ ] Make sure menu items are translated :) ! 😄 !
- [x] Click opens tab counter menu

### Browser fragment
- [x] Light mode, normal browsing mode
- [x] Light mode, private browsing mode
- [x] Dark mode, normal browsing mode
- [x] Dark mode, private browsing mode
- [ ] Private mode, close all open tabs
  - ✅   Close all private tabs so that it switches to normal mode with previously existing tabs
     - open any number of normal tabs (two digits or infinity show the padding the best). Open private tabs. "Close all" private tabs so that it switches back to normal mode - check tab number/infinity to make sure it's not cut off 
- [x] Normal mode, close all open tabs
- [ ] Tab counter spacing around numbers: single digit, double, infinity
- [x] Long press tab counter
  - [ ] Make sure menu items are translated
- [x] Click opens tab counter menu

**Review by csadilek — APPROVED** [review]
🚢
